pkg/agent/multicast/mcast_route.go

```diff
@@ -0,0 +1,429 @@
+		}
+
+	}
+	return nil
+}
+
+func (c *MRouteClient) externalInterfacesLeaveMgroup(mgroup net.IP) error {
+	for _, config := range c.multicastInterfaceConfigs {
+		addrIP := config.IPv4Addr.IP.To4()
+		groupIPStr := mgroup.To4()
+		err := syscall.SetsockoptIPMreq(c.sockFD, syscall.IPPROTO_IP, syscall.IP_DROP_MEMBERSHIP, &syscall.IPMreq{
+			Multiaddr: [4]byte{groupIPStr[0], groupIPStr[1], groupIPStr[2], groupIPStr[3]},
+			Interface: [4]byte{addrIP[0], addrIP[1], addrIP[2], addrIP[3]},
+		})
+		if err != nil {
+			return fmt.Errorf("cannot leave multicast group %s: %s", mgroup.String(), err.Error())
+		}
+	}
+	return nil
+}
+
+// parseIGMPMsg parses the kernel version into parsedIGMPMsg. Note there is a change in
+// the after linux 5.9 in the igmpmsg struct
+// https://github.com/torvalds/linux/commit/c8715a8e9f38906e73d6d78764216742db13ba0e
+// ,which causes difference in parsing vif.
+func (c *MRouteClient) parseIGMPMsg(msg []byte) (*parsedIGMPMsg, error) {
+	kernelVersion, err := runtime.GetKernelVersion()
+	if err != nil {
+		return nil, err
+	}
+	var vif uint16
+	if kernelVersion.Major >= 5 && kernelVersion.Minor > 9 {
+		vif = uint16(msg[10]) + (uint16(msg[11]) << uint16(8))
+	} else {
+		vif = uint16(msg[10])
+	}
+	return &parsedIGMPMsg{
+		VIF: vif,
+		Src: net.IPv4(msg[12], msg[13], msg[14], msg[15]),
+		Dst: net.IPv4(msg[16], msg[17], msg[18], msg[19]),
+	}, nil
+}
+
+// processIGMPNocacheMsg reads igmpMsg from the socket and configures
+// multicast route based on the message.
+// The outbound routing entries are for sending multicast traffic from pods
+// and the inbound routing entries are utilized to route multicast traffic from
+// outside cluster or external nodes to local pods.
+func (c *MRouteClient) processIGMPNocacheMsg() {
+	igmpmsg := <-c.igmpMsgChan
+	klog.V(2).InfoS("Received igmpmsg", "igmpmsg", igmpmsg)
+	bytemsg := []byte(igmpmsg)
+	if bytemsg[8] != multicastsyscall.IGMPMSG_NOCACHE {
+		klog.InfoS("Not a IGMPMSG_NOCACHE message", "igmpmsg", bytemsg)
+		return
+	}
+	msg, err := c.parseIGMPMsg(bytemsg)
+	if err != nil {
+		klog.ErrorS(err, "Error parsing kernel version")
+		return
+	}
+	gatewayVIF, _ := c.vifAllocator.getVIF(c.gatewayInterface)
+	if msg.VIF != uint16(gatewayVIF) {
+		err := c.addInboundMrouteEntry(msg.Src, msg.Dst, msg.VIF)
+		if err != nil {
+			klog.ErrorS(err, "Adding inbound mroute entry failed")
+		}
+	} else {
+		outboundInterfaces := make([]string, len(c.multicastInterfaceConfigs))
+		for i := range c.multicastInterfaceConfigs {
+			outboundInterfaces[i] = c.multicastInterfaceConfigs[i].Name
+		}
+		err = c.addOutboundMrouteEntry(msg.Src, msg.Dst, outboundInterfaces)
+		if err != nil {
+			klog.ErrorS(err, "Adding outbound mroute entry failed")
+		}
+	}
+}
+
+func (c *MRouteClient) runRouteClient(stopCh <-chan struct{}) {
+	klog.InfoS("Start running multicast routing daemon")
+	go func() {
+		for {
+			buf := make([]byte, 4096)
```

Comment:
Seems you missed my comment earlier. Repaste it:
> Define a constant for 4096. How we decided this value?

Reply:
After seaching in the web , I still don't have a  solid reason to specify a number here.

smcroute uses 128 bytes as buffer so I change it 128. I think the reason is 128 bytes buffer is enough to store an igmpmsg message.